### ISSUE TITLE
Builder pattern needed?

### ISSUE THREAD

**Issue by jruts** [torokmark/design_patterns_in_typescript#6]
Hi,

Since node/javascript does not have the problem that a builder pattern fixes, since you can just pass along an object literal, why would you want to apply all this overhead?

If we take a look at the example: https://github.com/torokmark/design_patterns_in_typescript/blob/master/builder/builder.ts  

We could just use an object literal instead:  

```javascript
// user.js
const create = args => {
  const options = Object.assign({}, defaultOptions, args); // Normalize the options
  return {
    name: options.name,
    age: options.age,
    ...
  }
}

module.exports.create = create;

// test.js
const user = require('user.js');

const myUser = user.create({
  name: 'a name',
  age: 30
});
```

This is similar to whatever you are trying to achieve with the builder pattern. The main difference here is that user.js is a lot smaller than the builder.ts example.   
Passing along an object literal, which is very common in node, is just as readable as with a builder, but with a lot less overhead or boilerplate.

So my question is: why would you do this?

**Comment by TomSeldon** [comment]
One use case I've come across (I'm going to try and dig up a really good blog post about this as well), is for when creating mock data in unit tests.

I used to do this very much like your above example.

e.g.

```javascript
let createUser;

beforeEach(() => {
    createUser = () => {
        return {
            username: 'bob.smith',
            email: 'bob.smith@aol.com',
            accountType: ACCOUNT_TYPES.DEVELOPER
       };
    };  
}));

// some spec

it('should do something', () => {
    const user = createUser();
});
```

This is all well and good, but what if we don't always want to test against a "developer" user. Well, we could do something like you suggested and pass in an object to override the parameters, or use individual parameters (e.g. `createUser => (username) => {}`).

This is fine, but if you often find yourself having to modify the default/generic object in order to obtain a useful object to test against, you can end up with lots of:

```javascript
const user = createUser({
    accountType: ACCOUNT_TYPES.ADMIN,
    username: 'some.user'
});

// ...

const user = createUser({
    accountType: ACCOUNT_TYPES.ADMIN,
    username: 'some.user'
});

// ...etc...
``` 

That seems fine in isolation, but if each of those definitions is really saying "create me an admin user", then that would be better abstracted to a `createAdminUser` function, which sets the admin defaults and allows overrides, similar to `createUser`.

**In a simple example like this, it's probably more overhead that it's worth to do anything more complicated than that.**

If, such as in a real, complicated application, you're needing to create users in a lot of different places (read: a lot of different tests), the above can quickly get repetitive and you can end up with lots of these types of functions littered through the test suite. In such a situation, having a generalised "userBuilder" can help reduce duplication through the tests.

Much of this can be applicable to non-test code, too, but I'm specifically focussing on that as it's what I found most useful.

Apologies for the slight incoherence, trying to type out this quickly during the downtime of waiting for something to compile...

**A really good blog post on this is:** http://davedevelopment.co.uk/2015/01/28/test-data-builders.html

The examples in that blog post are all PHP, but it's just as relevant here. Especially if you want to work with immutable data (though not a prerequisite to this being useful).

_________

**TL;DR:** For a very small or very simple use cases, it introduces overhead that almost certainly isn't worth it. But the same can be said of most abstractions. (One of?) the benefits comes with reducing complexity and repetitiveness in larger and more complicated use cases.

**Comment by jruts** [comment]
Thanks for the quick response!

I am kinda struggling to see where a builder would be easier for testing.  
Similar to where I have to `create` my object to create a user, you would also have to `build` your object with the builder.

so instead of:
```javacript
const user = createUser({
    accountType: ACCOUNT_TYPES.ADMIN,
    username: 'some.user'
});
```
you would need to do something like:
```javascript
const user = User(User.Builder('some.user').withAccountType(ACCOUNT_TYPES.ADMIN))
```

So I do not see the benefit of that in testing.  
If I needed a specific admin user for multiple tests, I would just create an admin user with the correct options.  
You would also need to create this admin user if you were to use the builder pattern.

So yes, I am still struggling to see 'why' :)

**Comment by msqaddura** [comment]
The main idea behind Software Design Patterns is that instancialization of an class object should be invisible  because instancialization the least you should worry about in a complex structure. 

let us say you wanted to add a log.. that contains all created user. if you were to use 
`const user = new User()`
then everytime you have to do 
`GlobalLog.addUser(user)`
then maybe you are only using it for debugging so
`if(GlobalDebug)GlobalLog.addUser(user);`

however if you use a structural pattern the one who instancialization it will do so and you wont have to do the very same trick for every file you have an instance in.

Now you will ask why dont I do that in the User Class... let me know how long it will take you to do it all over the whole system

**Comment by jruts** [comment]
I would never use a class in the first place when a simple object literal and simple functions can do the job. hence I do not need to create anything with the `new` keywords. Don't forget this is JavaScript. It does not have to be or look like Java

**Comment by msqaddura** [comment]
this example is all about typescript not plain javascript if you are moer into object literals then there is indeed no point in builders or any other design pattern... plain old javascript is not an OOP and this pattern is for OOP. so yes if you are more into object literals then this is not usable at all..you do not need it

**Comment by jruts** [comment]
@msqaddura , in javascript you have the choice to do OOP or functional. And I think you are missing the point. I was trying to find an answer to the following question:

> Why would you use a builder pattern when using plain javascript object literals gives you the same readability and the same functionality, but without any of the overhead or boilerplate.

I am not trying to attack any patterns here, I just do not think it belongs in every language. And since TypeScript is merely a superset of javascript, it is still javascript. 

And as [per definition of the builder pattern](https://en.wikipedia.org/wiki/Builder_pattern), the problem it is trying to solve is not present in javascript.

**Comment by msqaddura** [comment]
man I believe you are completely our of the context between Prototyping(apparently your style), Object Oriented(apparently not your style), Functional(everyone's style that doesnt really matter)... what am saying is just that for you and the way you look at javascript (over a typescript code) you do not need any pattern it wont help you anyway... unless you are hardcore Object Oriented with javascript/typescript then you can use a pattern but for you it is absolutely no point to bother about a pattern.

just because there is a pattern it doesnt really mean that we have to use it everywhere... it can only make things worse if we are forcing it into something that is not meant for it. 


I do use builders with typescript but ONLY when I have the urge and that was building a CMS you can imagine a an form tag <form> can take billions of options and CMS users should really bother about the way am handling it... something like {"type":"tag","family":"form","actions:[{beforeSubmit:"someFunctionILike"}]....

**Comment by msqaddura** [comment]
https://github.com/joelpalmer/Typescript--Builder-Pattern
please look at this example it should completely answer your original question that we drifted from.
            var myRequest = new scope.Utils.RequestBuilder("get").setScheme("http").setPath("mypath").setHost("localhost").setPort("9999").build();

without a builder it would be 
var myRequest = new Get("http","myPath","localHost",9999)
and cards constructor would be constructor(Scheme="http",path="myPath"......)

we can add destructing pattern over here for default values
constructor({Scheme="http",...})//ntices we wrapped it weith json tab 

so yes you can fiddle it down to literal objects but once change in superior constructor will have to be done to all constructors super() call... but with builder you would just need to change the method in the superior class... again you can find a way to simplify so it is always like this... you can always find your way out

**Comment by jruts** [comment]
@msqaddura you lost me at 'superior constructor'. 
> Favor composition over class inheritance

Never would I fall in the trap of extending, so never would I have the issue you just described.

And your example is again an example of how a Java person would write node. Just take a look at the 'request' module and see how it is done generally in node: https://github.com/request/request

```javascript
// You would do:
request.get({
  host: 'localhost',
  port: 9999,
  path: 'mypaeh'
}, callback)

// not
new Get("http","myPath","localHost",9999)
```

The only  reason why you are finding these issues is because you are not coming up with common nodejs/javascript examples. You find the problems by using javascript as if it was java.

**Comment by Kumjami** [comment]
you can still do it with something like 

```js
var options = {
   method: 'get',
   scheme: 'http',
   host: 'localhost',
   port: 9999
};
var request = createRequest(options);
```

and you don't need the 200 lines of code https://github.com/joelpalmer/Typescript--Builder-Pattern/blob/master/RequestBuilder/RequestBuilder.ts

you would need something like @jruts said here: https://github.com/torokmark/design_patterns_in_typescript/issues/6#issuecomment-284765974

**Comment by msqaddura** [comment]
do you know the difference between javascript and typescript? I think you are asking the question on a typescript page while you are waiting for the answer in a javascript page. If so, then i believe we are both getting each another differently...

**Comment by Kumjami** [comment]
typescript adds types to javascript, it doesn't mean you have to forget about javascript essence

**Comment by msqaddura** [comment]
I dont want to live in this planet anymore :D first read, then ask

**Comment by jruts** [comment]
Typescript is a superset of javascript. Which means plain javascript is valid typescript. The `benefit` that typescript gives you is nice IDE support like auto-completion and compile errors. You could easily write a definition for the object literal you want to pass along and it would be valid javascript with the typescript benefits.  

But let's make it very clear, adding typescript does not mean this language becomes Java and it does not mean you should treat it in the same way.   

By adding typescript for it's `benefits` to then add all the complexity of design patterns that you do not need kinda defeats the purpose for me.

**Comment by msqaddura** [comment]
![image](https://cloud.githubusercontent.com/assets/11808840/23699606/71e9a1dc-03f8-11e7-8763-a62bb956bd42.png)

add to types: modules & classes & interface

**Comment by msqaddura** [comment]
I recommend closing the issue as it is going more into argument than a solution.

**Comment by jruts** [comment]
@msqaddura really... all I want to do is open a discussion where we could all learn from, yet you fail to address anything I mentioned in my comment and post a picture of what AtScript is?

I do sense you are being very defensive but fail to come up with real examples or benefits. Even the examples you post in pure javascript clearly indicate that you have no great experience with pure javascript at all.

All I was hoping for was a conclusion to my question as to why would I ever use a builder pattern as a javascript developer. There is no need for childish discussion.

**Comment by msqaddura** [comment]
that is what am saying, no need for childish discussion

**Comment by Phmager** [comment]
Even though the last comment was quite long ago, but I just stumbled across this question.
In my opinion it also adds a lot of value in the javascript/typescript world. Of course it is dependent on the use case if it is usefull or not. 

Here are the advantages:
### Decouples the model of the creation of the model. 
If your model changes often, you don't need to change as many places. -> Clearer git history on business files, less conflicts due to refactorings.

### Radability is improved
Readability is especially improved, if you have a huge object with a lot of properties and when you create a lot of elements through the builder:

If you take a look at the example, that this dense code would need for **each** of the method chains at least 1-2 more lines, it is harder too read.
![image](https://user-images.githubusercontent.com/9350951/51969842-b18b9800-2475-11e9-838b-64f52897bd60.png)

### Context
A real builder can have multiple sub builders, that depend on the called method. In the example above for the first call you have different possibilities of initialization, which will change some fundamental things.

### Extensibility / Transformations
The builder can add very usefull utilities to make the usage of the model a lot easier. A junior dev might not know dependencies between different properties of the model. But they can be encapsulated into a nice api using a builder.

They can set multiple properties (also with good default values specially for these calls)
![image](https://user-images.githubusercontent.com/9350951/51970169-79d12000-2476-11e9-9435-a08ae7e6f3c3.png)

They can set pre defined values
![image](https://user-images.githubusercontent.com/9350951/51970190-86ee0f00-2476-11e9-9a25-7af9eefdf437.png)

They can even extend the usages, by allowing transformation of inputs. This way the model does not need to be extended, as long as we can map to the result in some way:

![image](https://user-images.githubusercontent.com/9350951/51970234-a84efb00-2476-11e9-90c9-a31635d94fa5.png)

---

I hope this makes it clear, that there are indeed use cases for such a pattern.

Btw. jQuery is a huge builder pattern. (https://www.w3schools.com/jquery/jquery_chaining.asp)
